README.md
# draw2Sample
the interface for graphical sampling in order to generate values following an empirical distribution, with img/canvas tag on Firefox.  
https://github.com/YujiSODE/draw2Sample

>Copyright (c) 2016 Yuji SODE \<yuji.sode@gmail.com\>  
>This software is released under the MIT License.  
>See LICENSE or http://opensource.org/licenses/mit-license.php
______

##Concept
![Concept of graphical sampling:draw2Sample_Fig.png](./draw2Sample_Fig.png)  
__Figure 1. Concept of graphical sampling.__ _f_ (_x_) is a graph drawn on graphical data (_W_ x _H_ px).

The sample (_Sp_) as a result of graphical sampling with graphical data: _W_ x _H_ px (Figure 1) can be expressed as follows:  
let graphical heights of _f_ (_x1_) and _f_ (_x2_) be _a1_ px and _a2_ px respectively,  
(1)   _Sp_ = _x0_, ..., _x1_, ..., _x1_, _x2_, ..., _x2_, ..., _xn_  
and  
(2)   _Sp_ = _x0_ \* _a0_, _x1_ \* _a1_, _x2_ \* _a2_, ..., _xn_ \* _an_  
RGBA color value in a pixel is available in order to recognize the shape of the graph.  
The alpha value is used in this program; a pixel where alpha > 0 is regarded as graph.

##Script
* draw2Sample.js
* \[v1.2+\] pValue.js

##How to generating values following a given distribution
1. __"Sample (_Sp_)":__ ["Rnd"](#output-data) in the result output of "draw2Sample.js"  
2. __Generating target values;__ the target values are obtained with bootstrap Method (Efron,1979) via _Sp_.
3. __\(Optional \[v1.2+\]\) Estimating _p_-value;__ _p_-value is estimated with "pValue.js" via _Sp_.

##How to use
* call "draw2Sample()" in a html file with img/canvas tag.
* \(Optional \[v1.2+\]\) call "_pValue()" of "pValue.js", in order to estimate _p_-value; (see [__Estimating _p_-value__](#v12-estimating-p-value) for details of prameters).

1. __Selecting some target images__  
   __Target tag;__ here target images can be selected.  

2. __Draw graph__  
   __Drawing;__ a graph can be drawn on the newly created canvas tag freehand or with script (see [__"Example with the standard normal distribution"__](#example-with-the-standard-normal-distribution)).  
   "canvas id": the id of canvas tag to draw graph.  
   "Size": size of the selecting canvas tag, expressed with Width (W) and Height (H).

3. __Sampling__  
   The sampling from the given graph (2.) is run with __"Run sampling" button__, and the result output is shown in __"Result"__.  

   ####_Inputs and buttons_  
   * __"Target width":__ the target sampling area, expressed with left side x coordinate (x0) and width (w), shown in red.
   * __"Sampling interval";__ it sets how many times the given graph (2.) is sampled in a given target area, starting with x0.
   * __"Range of values":__ the true x-coordinate values in the target sampling area, expressed with left side (v0) and right side (v).
   * __"Clear image" button;__ it clears the selected image (1.), which is shown under the drawing layer.
   * __"Show sampling line" button;__ it shows "Sampling interval" as vertical lines.
   * __"Clear sampling line" button;__ it clears "Sampling interval" shown by __"Show sampling line" button__.
   * __"Clear drawing" button;__ it clears drawn graph (2.).
   * __"Run sampling" button;__ it runs sampling and outputs results into __"Result"__.
   * __"Close" button;__ it closes this interface.
   * __"Result";__ where results are output, and additional comments are also available here.
   * __"Clear result" button;__ it clears output results in __"Result"__.
   * __"Email address":__ email address used outputting __"Result"__ as email format.
   * __"Output as email" button;__ it saves the __"Result"__ as email to given address.

   ####_Output data_  
   * dataLog: csv formatted values expressed as _n_@_y_ for a _n_-th sampling result: _y_ with top left corner as origin.
   * x@f(x): csv formatted values expressed as _x_@_f_ (_x_) for a value of _f_ (_x_) at _x_ with bottom left as origin.
   * __Rnd:__ csv formatted values estimated as results of a sampling.

______
##\[v1.2+\] Estimating _p_-value
####__"pValue.js"__
* Probability estimator with given numerical data and bootstrap Method (Efron,1979) on Firefox.  
`_pValue(data,x,sampleSize,simulation)`  
`/*`  
`*===<parameter>===`  
`* data: numerical data as csv formatted text e.g., '1,2,3,...'`  
`*===<optional parameters>===`  
`* x: numerical value; the upper 25% of a given values as default`  
`* sampleSize: numerical positive integer; 100 as default`  
`* simulation: numerical positive integer; 10 as default`  
`*/`  
_p_-value is estimated as probability on v-axis: _P_(`x`) = _P_(_v_ >= `x`).  
The estimation is based on resampled data with size (`sampleSize`) for n-time (`simulation`) simulations.

####Examples
Script 1: `_pValue('1,2');` the result: `{"p":0.504,"x":1.75,"sampleSize":100,"simulation":10}`  
Script 2: `_pValue('1,2,3');` the result: `{"p":0.323,"x":2.5,"sampleSize":100,"simulation":10}`  
Script 3: `_pValue('1,2,3',1.1);` the result: `{"p":0.676,"x":1.1,"sampleSize":100,"simulation":10}`  
Script 4: `_pValue('1,2,3',2,10,1);` the result: `{"p":0.6,"x":2,"sampleSize":10,"simulation":1}`

______
##Example with the standard normal distribution
###Script used for drawing graph
* stdNormDist100pt.js

###Sampling parameters
`/*Fri_Sep_09_2016_17:01:16_GMT+0900_(JST),Sampling interval:20,Size: W x H = 400 x 400 px*/`

![Sampling example with the standard normal distribution by script:stdNormDist_scriptOutputAll.jpg](./stdNormDist_scriptOutputAll.jpg)  
__Figure 2. Sampling example with the standard normal distribution by script.__ Blue and red lines show a graph  
of the standard normal distribution by script and a recognized distribution respectively. Vertical lines show where  
sampled from blue graph.

###1-sample t-test
The sampled size by "draw2Sample.js": `<Sample size:2527>`

1. `[Resampled size:100;Given mean:0;Simulation:100times;Significance level:0.025]`  
   Result: `Rejection rate:0`
2. `[Resampled size:10;Given mean:0.0;Simulation:10000times;Significance level:0.025]`  
   Result: `Rejection rate:0.0163`
3. `[Resampled size:20;Given mean:0;Simulation:10000times;Significance level:0.025]`  
   Result: `Rejection rate:0.0146`
   
###_p_-value estimation
The sampled size by "draw2Sample.js": `<Sample size:2527>`  
_p_-value was estimated as probability on v-axis: _P_(`x`) = _P_(_v_ >= `x`).

1. _P_(the upper 25% of given data): `{"p":0.035,"x":1.7000000000000002,"sampleSize":100,"simulation":10}`
2. _P_(0): `{"p":0.554,"x":0,"sampleSize":100,"simulation":10}`
3. _P_(-1.96): `{"p":0.966,"x":-1.96,"sampleSize":100,"simulation":10}`
4. _P_(1.96): `{"p":0.028,"x":1.96,"sampleSize":100,"simulation":10}`

______
##Reference
* Efron, B. 1979. Bootstrap Methods: Another Look at the Jackknife. Ann. Statist. vol. 7, no. 1, p. 1-26.

##\[1.2+\] Library list
* bootstrapEst-2.1/bootstrapMdl.js (Yuji SODE,2016): the MIT License; https://github.com/YujiSODE/bootstrapEst


CHANGELOG.md
# Change Log

## [1.2.1] - 2016-11-10

## Changed
- [README.md] line 62-63: `   * __dataLog:__ csv formatted values expressed as _n_@_y_ for a _n_-th sampling result: _y_ with top left corner as origin.`  
  `   * __*x*@*f(x)*:__ csv formatted values expressed as _x_@_f_ (_x_) for a value of _f_ (_x_) at _x_ with bottom left as origin.`

## Added
- [README.md] line 124-125: `##\[1.2+\] Library list`  
  `* bootstrapEst-2.1/bootstrapMdl.js (Yuji SODE,2016): the MIT License; https://github.com/YujiSODE/bootstrapEst`

## [1.2] - 2016-10-29

## Added
- [README.md] line 115-118: `1. _P_(the upper 25% of given data):` `{"p":0.035,"x":1.7000000000000002,"sampleSize":100,"simulation":10}`  
  `2. _P_(0):` `{"p":0.554,"x":0,"sampleSize":100,"simulation":10}`  
  `3. _P_(-1.96):` `{"p":0.966,"x":-1.96,"sampleSize":100,"simulation":10}`  
  `4. _P_(1.96):` `{"p":0.028,"x":1.96,"sampleSize":100,"simulation":10}`
- [README.md] line 111-113: `###_p_-value estimation`  
  `The sampled size by "draw2Sample.js":` `<Sample size:2527>`  
  `_p_-value was estimated as probability on v-axis: _P_(`x`) = _P_(_v_ >= `x`).`

1. _P_(the upper 25% of given data): `{"p":0.035,"x":1.7000000000000002,"sampleSize":100,"simulation":10}`
2. _P_(0): `{"p":0.554,"x":0,"sampleSize":100,"simulation":10}`
3. _P_(-1.96): `{"p":0.966,"x":-1.96,"sampleSize":100,"simulation":10}`
4. _P_(1.96): `{"p":0.028,"x":1.96,"sampleSize":100,"simulation":10}`

## Changed
- [README.md] line 79: `_p_-value is estimated as probability on v-axis: _P_(`x`) = _P_(_v_ >= `x`).`
- [README.md] line 67: `##\[v1.2+\] Estimating _p_-value`
- [README.md] line 33: `* \(Optional \[v1.2+\]\) call "_pValue()" of "pValue.js", in order to estimate _p_-value; (see [__Estimating _p_-value__](#v12-estimating-p-value) for details of prameters).`
- [README.md] line 29: `3. __\(Optional \[v1.2+\]\) Estimating _p_-value;__ _p_-value is estimated with "pValue.js" via _Sp_.`

## Added
- [README.md] line 82-86: `####Examples`  
  Script 1: `_pValue('1,2');` the result: `{"p":0.504,"x":1.75,"sampleSize":100,"simulation":10}`  
  Script 2: `_pValue('1,2,3');` the result: `{"p":0.323,"x":2.5,"sampleSize":100,"simulation":10}`  
  Script 3: `_pValue('1,2,3',1.1);` the result: `{"p":0.676,"x":1.1,"sampleSize":100,"simulation":10}`  
  Script 4: `_pValue('1,2,3',2,10,1);` the result: `{"p":0.6,"x":2,"sampleSize":10,"simulation":1}`
- [README.md] line 79-80: `p-value is estimated as probability: _P_(`x`) = _P_(_v_ >= `x`).`  
  The estimation is based on resampled data with size (`sampleSize`) for n-time (`simulation`) simulations.
- [README.md] line 70-78: `_pValue(data,x,sampleSize,simulation)`  
  `/*`  
  `*===<parameter>===`  
  `* data: numerical data as csv formatted text e.g., '1,2,3,...'`  
  `*===<optional parameters>===`  
  `* x: numerical value; the upper 25% of a given values as default`  
  `* sampleSize: numerical positive integer; 100 as default`  
  `* simulation: numerical positive integer; 10 as default`  
  `*/`
- [README.md] line 68-69: `####__"pValue.js"__`  
  `* Probability estimator with given numerical data and bootstrap Method (Efron,1979) on Firefox.`

## Changed
- [README.md] line 33: `* \(Optional \[v1.2+\]\) call "_pValue()" of "pValue.js", in order to estimate p-value; (see [__Estimating p-value__](#v12-estimating-p-value) for details of prameters).`

## Added
- [README.md] line 66-67: `______`  
  `##\[v1.2+\] Estimating p-value`
- [README.md] line 33: `* \(Optional \[v1.2+\]\) call "_pValue()" of "pValue.js", in order to estimate p-value; (see [__Estimating p-value__]() for details of prameters).`
- [README.md] line 29: `3. __\(Optional \[v1.2+\]\) Estimating p-value;__ p-value is estimated with "pValue.js" via _Sp_.`
- [README.md] line 24: `* \[v1.2+\] pValue.js`

## [1.2] - 2016-10-28

## Changed
- [pValue.js] line 21: `  if(+x!=0){x=(!x||!/^[+-]?[0-9]+(?:\.[0-9]+)?$/.test(+x))?'up25(d)':x;}else{x=0;}`

## [1.2] - 2016-10-27

## Changed
- [pValue.js] line 48-50: `  W.addEventListener('message',function(e){`  
  `    R.p=e.data[0],R.x=e.data[1],R.sampleSize=sampleSize,R.simulation=simulation;`  
  `    slf.document.getElementsByTagName('body')[0].innerHTML+='<textarea>'+slf.JSON.stringify(R)+'</textarea>',R=null;`  
  `  },true);`
- [pValue.js] line 19: `  var slf=window,size=0,b,U,W,spt='',R={p:0,x:0,sampleSize:0,simulation:0};`
- [pValue.js] line 18: `  function _pValue(data,x,sampleSize,simulation){`

## [1.1beta] - 2016-10-26

## Added
- [pValue.js] line 9: `//probability estimator with given numerical data and bootstrap Method (Efron,1979) on Firefox.`

## [1.1beta] - 2016-10-25

## Changed
- [pValue.js] line 29-34: added script using modified version of bootstrapEst-2.1/bootstrapMdl.js;

## Added
- added files: pValue.js

## [1.1beta] - 2016-10-24

## Removed
- [bootstrapEst-2.1/bootstrapMdl.js] line 6-13: removed

## Changed
- [bootstrapEst-2.1/bootstrapMdl.js] line 14: `'var bts=function(A,n){var r=[],rdInt=0;for(var i=0;i<n;i+=1){rdInt=Math.floor(A.length*Math.random());r.push(A[rdInt]);}return r;},'`
- [bootstrapEst-2.1/bootstrapMdl.js] line 1-10: `/*bootstrapEst-2.1/bootstrapMdl.js (Yuji SODE,2016)`  
  `* This software is released under the MIT License. See http://opensource.org/licenses/mit-license.php`  
  `* the simple interface for estimation of a value of statistic, with bootstrap Method (Efron,1979).`  
  `* reference: Efron, B. 1979. Bootstrap Methods: Another Look at the Jackknife. Ann. Statist. vol. 7, no. 1, p. 1-26.`  
  `*/`


stdNormDist100pt.js
/*stdNormDist100pt.js
* copyright (c) 2016 Yuji SODE <yuji.sode@gmail.com>
*/
//it draws 100 points: [-4,+4] graph data of the standard normal distribution.
//data was calculated with gnuplot version 4.6 patchlevel 4.
function _stdNormDist100pt(){
  var slf=this.window,cvsId,c,d,n=0,i=1,w0=0,h0=0,dw=0,dh=0;
  cvsId=slf.prompt('canvas id?');
  c=slf.document.getElementById(cvsId).getContext('2d');
  if(!c){throw 'the id/canvas tag is undefined';}
  w0=c.canvas.width,h0=c.canvas.height;
  dw=w0*0.01,dh=h0/0.3989422804014327;
  d=xy(),n=d.length;
  c.clearRect(0,0,w0,h0);
  c.strokeStyle='rgba(0,0,255,1)',c.lineWidth=1.0;
  //drawing graph
  c.beginPath(),c.moveTo(0,h0-d[0][1]*dh);
  while(i<n){c.lineTo(i*dw,h0-d[i][1]*dh),i+=1;}
  c.stroke();
//====== data calculated with GNUPLOT Version 4.6 patchlevel 4 ======
//it returns 100 points: [-4,+4] graph data of the standard normal distribution.
    function xy(){
      return [
      [-4,0.00013383],[-3.91919,0.000184295],[-3.83838,0.000252138],[-3.75758,0.00034271],[-3.67677,0.000462785],[-3.59596,0.000620862],
      [-3.51515,0.000827515],[-3.43434,0.00109577],[-3.35354,0.00144155],[-3.27273,0.00188409],[-3.19192,0.00244646],[-3.11111,0.00315602],
      [-3.0303,0.00404487],[-2.94949,0.00515031],[-2.86869,0.00651518],[-2.78788,0.00818811],[-2.70707,0.0102236],[-2.62626,0.0126821],
      [-2.54545,0.0156293],[-2.46465,0.0191361],[-2.38384,0.0232772],[-2.30303,0.0281302],[-2.22222,0.0337737],[-2.14141,0.0402854],
      [-2.06061,0.0477399],[-1.9798,0.0562056],[-1.89899,0.0657418],[-1.81818,0.0763955],[-1.73737,0.0881979],[-1.65657,0.101161],
      [-1.57576,0.115274],[-1.49495,0.130501],[-1.41414,0.146778],[-1.33333,0.16401],[-1.25253,0.182073],[-1.17172,0.200809],
      [-1.09091,0.220033],[-1.0101,0.239527],[-0.929293,0.259051],[-0.848485,0.278343],[-0.767677,0.297125],[-0.686869,0.31511],
      [-0.606061,0.332009],[-0.525253,0.347537],[-0.444444,0.361424],[-0.363636,0.373419],[-0.282828,0.383301],[-0.20202,0.390884],
      [-0.121212,0.396022],[-0.040404,0.398617],[0.040404,0.398617],[0.121212,0.396022],[0.20202,0.390884],[0.282828,0.383301],
      [0.363636,0.373419],[0.444444,0.361424],[0.525253,0.347537],[0.606061,0.332009],[0.686869,0.31511],[0.767677,0.297125],
      [0.848485,0.278343],[0.929293,0.259051],[1.0101,0.239527],[1.09091,0.220033],[1.17172,0.200809],[1.25253,0.182073],
      [1.33333,0.16401],[1.41414,0.146778],[1.49495,0.130501],[1.57576,0.115274],[1.65657,0.101161],[1.73737,0.0881979],
      [1.81818,0.0763955],[1.89899,0.0657418],[1.9798,0.0562056],[2.06061,0.0477399],[2.14141,0.0402854],[2.22222,0.0337737],
      [2.30303,0.0281302],[2.38384,0.0232772],[2.46465,0.0191361],[2.54545,0.0156293],[2.62626,0.0126821],[2.70707,0.0102236],
      [2.78788,0.00818811],[2.86869,0.00651518],[2.94949,0.00515031],[3.0303,0.00404487],[3.11111,0.00315602],[3.19192,0.00244646],
      [3.27273,0.00188409],[3.35354,0.00144155],[3.43434,0.00109577],[3.51515,0.000827515],[3.59596,0.000620862],[3.67677,0.000462785],
      [3.75758,0.00034271],[3.83838,0.000252138],[3.91919,0.000184295],[4,0.00013383]
      ];}
}


pValue.js
/*draw2Sample
* pValue.js
*
*    Copyright (c) 2016 Yuji SODE <yuji.sode@gmail.com>
*
*    This software is released under the MIT License.
*    See LICENSE or http://opensource.org/licenses/mit-license.php
*/
//probability estimator with given numerical data and bootstrap Method (Efron,1979) on Firefox.
/*
*===<parameter>===
* data: numerical data as csv formatted text e.g., '1,2,3,...'
*===<optional parameters>===
* x: numerical value; the upper 25% of a given values as default
* sampleSize: numerical positive integer; 100 as default
* simulation: numerical positive integer; 10 as default
*/
function _pValue(data,x,sampleSize,simulation){
  var slf=window,size=0,b,U,W,spt='',R={p:0,x:0,sampleSize:0,simulation:0};
  //if x is numerical x=x, otherwise x='up25(d)'; 'up25(d)' is the upper 25% of a given values
  if(+x!=0){x=(!x||!/^[+-]?[0-9]+(?:\.[0-9]+)?$/.test(+x))?'up25(d)':x;}else{x=0;}
  //smpleSize: positive integer
  if(!sampleSize){sampleSize=100;}else if(!/^[+]?[1-9][0-9]*$/.test(sampleSize)){throw new Error('sampleSize: invalid value');}
  //simulation: positive integer
  if(!simulation){simulation=10;}else if(!/^[+]?[1-9][0-9]*$/.test(simulation)){throw new Error('simulation: invalid value');}
  /*script for worker*/
  spt=[
    /*this method estimates max and min value of an array; returned value is [max,min]*/
    'this.Array.prototype.mx=function(){var A=this.sort(function(a,b){return b-a;});return [A[0],A[A.length-1]];};',
    /*bootstrapEst-2.1/bootstrapMdl.js (Yuji SODE,2016)
    * This software is released under the MIT License. See http://opensource.org/licenses/mit-license.php
    * the simple interface for estimation of a value of statistic, with bootstrap Method (Efron,1979).
    * reference: Efron, B. 1979. Bootstrap Methods: Another Look at the Jackknife. Ann. Statist. vol. 7, no. 1, p. 1-26.
    */
    'var bts=function(A,n){var r=[],rdInt=0;for(var i=0;i<n;i+=1){rdInt=Math.floor(A.length*Math.random());r.push(A[rdInt]);}return r;},',
    /*it returns a value for the upper 25% of the given data range*/
    'up25=function(A){var v=A.mx(),d=(v[0]-v[1])/4;return v[1]+d*3;},',
    'd=['+data+'],_x='+x+',B,n=0,m='+simulation+',R=0,i=0,j=0;',
    'this.addEventListener(\'message\',function(){while(i<m){j=0;B=bts(d,'+sampleSize+'),n=B.length;while(j<n){R+=B[j]<_x?0:1,j+=1;}i+=1;}',
    /*[estimated p-value,value of _x] is posted*/
    'this.postMessage([R/'+sampleSize*simulation+',_x]),d=_x=B=n=m=R=null;this.close();},true);'
  ].join('');
  //generation of worker
  b=new Blob([spt],{type:'text/javascript'});
  U=slf.URL.createObjectURL(b);
  W=new Worker(U),slf.URL.revokeObjectURL(U),b=null;
  //eventlisteners with worker
  W.addEventListener('message',function(e){
    R.p=e.data[0],R.x=e.data[1],R.sampleSize=sampleSize,R.simulation=simulation;
    slf.document.getElementsByTagName('body')[0].innerHTML+='<textarea>'+slf.JSON.stringify(R)+'</textarea>',R=null;
  },true);
  //error in worker
  W.addEventListener('error',function(e){console.log(e.message),W.terminate();},true);
  W.postMessage('Run');
};
